### DISCUSSION TITLE
L3D Editor exports not well formatted XML with missing Declaration

### DISCUSSION THREAD

**Discussion by holg** [globallightingdata/l3d#15] [General]
XML Declaration is missing in the structure.xml of L3D Editor saved files:
The XML declaration is optional but highly recommended, especially for documents that use encodings other than UTF-8 or UTF-16. If the XML document is encoded in UTF-8 (the default for many systems)
As well this file has \r\n line endings, which is a bit misleading, as this is typical for Windows, where UTF-8 (without BOM) is not usual.
So the XML Declaration at least would lead to the right direction...
Another strict point:
You export the missing Name as 
<Name/>
Since Name is optional you should according to standards omit it completely IMHO.
Another point is the closing of the Tag:
<FormatVersion major="0" minor="11"/>
should be 
<FormatVersion major="0" minor="11" />
Both are valid XML, but W3C recommends the last one

**Comment by Kurpanik** [comment]
The  XML standard does provides very strict guidance on how to handle the situation when the encoding is missing. According to the 1.0 specification, if the encoding declaration is missing, you simply have to assume UTF-8. That's all. 

Ignoring this standard and making somewhat strange assumptions about the encoding based on the file line endings and historical practices under Windows, is something I really do not recommend.

**Reply by holg** [reply]
Thanks for the quick response, i really appreciate.

Well i agree 100%.
This is why i. outlined this.
Right now it needs quite some XML sanitization.
It is the XML from the customers, which contain UTF with Bom and such, most likely bcs of copy paste on Windows.
So i am doing some Open Source L3D parser alike GLDF i did.
Doing tests with existing files from various sources.
Since you as well does not recommend,  is why i proposed not to forget the XML Declaration.

**Comment by Kurpanik** [comment]
> **It is the XML from the customers**, which contain UTF with Bom and such, most likely bcs of copy paste on Windows. Right now it needs quite some XML sanitization.

What I don’t understand: why does the L3D editor then need to be adjusted? Since you titled this discussion _L3D Editor exports not well formatted XML_. The L3D Editor doesn't provide any declaration, since it's encoding correctly to UTF-8. And correctly to the XML standard.

It sounds to me like you should be getting in touch with your customers instead?

**Reply by holg** [reply]
I mention it, because i think every little helps.
Being valid must not mean it is best practice,
which would behave the xml declaration, which can help linters,
and to omit empty fields.
I absolutely appreciate the  l3d editor, and as such wanted to help to improve it.

**Reply by Kurpanik** [reply]
Ok, I got your point and create an issue in our Backlog for it. However, it will be of rather low priority, since it's creating the XML according to the standard already. But yeah, no best practice as you say. Something we can add "on occasion" :). Thanks for the discussion, I appreciate it.
